File: boj/Implementation/BOJ_2504.swift
/*
 백준 2504 괄호의 값
 21.08.18
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

let input = readLine()!
let bracket = ["(", ")", "[", "]"]
var stack = [String]()

func addNums(_ n: Int) {
  var num = n

  while !stack.isEmpty && !bracket.contains(stack.last!) {
    num += Int(stack.removeLast())!
  }

  stack.append(String(num))
}

func multiNums(_ n: Int, _ pair: String) -> Bool {
  var num = n

  while !stack.isEmpty && !bracket.contains(stack.last!) {
    num *= Int(stack.removeLast())!
  }

  if stack.isEmpty || stack.last! != pair {
    return false
  } else {
    stack.removeLast()
  }

  if !stack.isEmpty && !bracket.contains(stack.last!) {
    num += Int(stack.removeLast())!
  }

  stack.append(String(num))
  return true
}

for now in input {
  if now == "(" || now == "[" {
    stack.append(String(now))
  } else {
    if stack.isEmpty {
      break
    }

    if now == ")" {
      if stack.last! == "(" {
        stack.removeLast()
        addNums(2)
      } else if stack.last! != "[" {
        if !multiNums(2, "(") {
          break
        }
      } else {
        break
      }

    } else if now == "]" {
      if stack.last! == "[" {
        stack.removeLast()
        addNums(3)
      } else if stack.last! != "(" {
        if !multiNums(3, "[") {
          break
        }
      } else {
        break
      }
    }
  }
}

if stack.isEmpty || stack.count > 1 || bracket.contains(stack[0]) {
  print(0)
} else {
  print(stack[0])
}

File: boj/Implementation/BOJ_2504.cpp
/*
 백준 2504 괄호의 값
 21.07.08
 https://github.com/skyqnaqna/algorithm_study
 */

#include <bits/stdc++.h>

using namespace std;

#define INF 1e9
#define endl "\n"

typedef long long ll;
typedef double dd;
typedef pair<int, int> pii;
typedef pair<ll, ll> pll;

stack<string>st;

int main() {
    ios::sync_with_stdio(0); cin.tie(0);

    string s; cin >> s;

    for (auto &c : s) {
        string temp;
        if (c == '(' || c == '[') st.push(temp+=c);
        else if (st.empty()) {
            cout << 0 << endl;
            return 0;
        } else if (c == ')') {
            if (st.top() == "[") {
                cout << 0 << endl;
                return 0;
            }

            int num = 2;

            if (st.top() == "(") {
                st.pop();
            } else {
                num *= stoi(st.top());
                st.pop();

                if (st.empty() || st.top() != "(") {
                    cout << 0 << endl;
                    return 0;
                }

                if (!st.empty() && st.top() == "(") st.pop();
            }

            if (!st.empty() && st.top() != "(" && st.top() != "[") {
                num += stoi(st.top());
                st.pop();
            }

            st.push(to_string(num));

        } else if (c == ']') {
            if (st.top() == "(") {
                cout << 0 << endl;
                return 0;
            }

            int num = 3;

            if (st.top() == "[") {
                st.pop();
            } else {
                num *= stoi(st.top());
                st.pop();

                if (st.empty() || st.top() != "[") {
                    cout << 0 << endl;
                    return 0;
                }
                if (!st.empty() && st.top() == "[") st.pop();
            }

            if (!st.empty() && st.top() != "(" && st.top() != "[") {
                num += stoi(st.top());
                st.pop();
            }

            st.push(to_string(num));

        }
    }


    if (st.empty() || st.size() > 1) cout << 0 << endl;
    else cout << st.top() << endl;

    return 0;
}